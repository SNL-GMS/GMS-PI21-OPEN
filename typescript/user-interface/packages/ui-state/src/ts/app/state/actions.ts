import { createAction } from '@reduxjs/toolkit';

/**
 * Action used to reset the application state.
 */
export const reset = createAction('reset');
