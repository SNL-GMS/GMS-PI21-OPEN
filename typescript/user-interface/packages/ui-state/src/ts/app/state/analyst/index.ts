export * from './analyst-slice';
export * as AnalystWorkspaceOperations from './operations';
export * from './selectors';
export * as AnalystWorkspaceTypes from './types';
