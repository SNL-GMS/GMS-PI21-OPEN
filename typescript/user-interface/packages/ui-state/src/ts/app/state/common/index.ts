export * from './common-slice';
export * from './operations';
export * from './types';
