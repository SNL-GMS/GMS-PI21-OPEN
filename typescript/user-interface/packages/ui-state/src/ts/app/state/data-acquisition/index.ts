export * from './data-acquisition-slice';
export * from './types';
