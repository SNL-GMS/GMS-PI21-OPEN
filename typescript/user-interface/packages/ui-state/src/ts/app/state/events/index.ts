export * from './events-slice';
export * from './types';
