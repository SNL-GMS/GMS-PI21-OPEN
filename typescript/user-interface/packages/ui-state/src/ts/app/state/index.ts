export * as AppActions from './actions';
export * from './analyst';
export * from './common';
export * from './data-acquisition';
export * from './events';
export * from './keypress';
export * from './map';
export * as AppOperations from './operations';
export * from './signal-detections';
export * from './station-properties-configuration';
export * from './system-message';
export * from './user-session';
export * from './waveform';
export * from './workflow';
