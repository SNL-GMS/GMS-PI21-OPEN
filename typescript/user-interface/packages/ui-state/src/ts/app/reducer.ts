import { combineReducers } from '@reduxjs/toolkit';

import {
  eventManagerApiSlice,
  processingConfigurationApiSlice,
  processingStationApiSlice,
  signalEnhancementConfigurationApiSlice,
  sohAceiApiSlice,
  ssamControlApiSlice,
  stationDefinitionSlice,
  systemEventGatewayApiSlice,
  systemMessageDefinitionApiSlice,
  userManagerApiSlice,
  workflowApiSlice
} from './api';
import { dataSlice } from './api/data/data-slice';
import { reducer as appReducer } from './state/reducer';

export const reducer = combineReducers({
  // application api (queries)
  [systemEventGatewayApiSlice.reducerPath]: systemEventGatewayApiSlice.reducer,
  [eventManagerApiSlice.reducerPath]: eventManagerApiSlice.reducer,
  [processingConfigurationApiSlice.reducerPath]: processingConfigurationApiSlice.reducer,
  [processingStationApiSlice.reducerPath]: processingStationApiSlice.reducer,
  [signalEnhancementConfigurationApiSlice.reducerPath]:
    signalEnhancementConfigurationApiSlice.reducer,
  [ssamControlApiSlice.reducerPath]: ssamControlApiSlice.reducer,
  [sohAceiApiSlice.reducerPath]: sohAceiApiSlice.reducer,
  [stationDefinitionSlice.reducerPath]: stationDefinitionSlice.reducer,
  [systemMessageDefinitionApiSlice.reducerPath]: systemMessageDefinitionApiSlice.reducer,
  [userManagerApiSlice.reducerPath]: userManagerApiSlice.reducer,
  [workflowApiSlice.reducerPath]: workflowApiSlice.reducer,
  // data state
  [dataSlice.name]: dataSlice.reducer,
  // application state
  app: appReducer
});
