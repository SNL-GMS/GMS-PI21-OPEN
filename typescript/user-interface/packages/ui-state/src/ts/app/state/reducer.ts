import type { CombinedState } from '@reduxjs/toolkit';
import { combineReducers, createReducer } from '@reduxjs/toolkit';
import type { ReducerWithInitialState } from '@reduxjs/toolkit/dist/createReducer';

import {
  analystInitialState,
  analystSlice,
  commonInitialState,
  commonSlice,
  dataAcquisitionInitialState,
  dataAcquisitionSlice,
  waveformInitialState,
  waveformSlice,
  workflowInitialState,
  workflowSlice
} from '.';
import { reset } from './actions';
import type { AnalystState } from './analyst/types';
import type { CommonState } from './common/types';
import type { DataAcquisitionState } from './data-acquisition/types';
import type { EventsState } from './events';
import { eventsInitialState, eventsSlice } from './events';
import type { MapState } from './map';
import { mapInitialState, mapSlice } from './map';
import type { SignalDetectionsState } from './signal-detections';
import { signalDetectionsInitialState, signalDetectionsSlice } from './signal-detections';
import {
  stationPropertiesConfigurationInitialState,
  stationPropertiesConfigurationSlice
} from './station-properties-configuration';
import type { StationPropertiesConfigurationState } from './station-properties-configuration/types';
import {
  systemMessageInitialState,
  systemMessageSlice
} from './system-message/system-message-slice';
import type { SystemMessageState } from './system-message/types';
import type { UserSessionState } from './user-session/types';
import { userSessionInitialState, userSessionSlice } from './user-session/user-session-slice';
import type { WaveformState } from './waveform/types';
import type { WorkflowState } from './workflow/types';

// combine all reducers
const reducers = combineReducers<{
  analyst: AnalystState;
  stationPropertiesConfiguration: StationPropertiesConfigurationState;
  common: CommonState;
  dataAcquisition: DataAcquisitionState;
  events: EventsState;
  map: MapState;
  signalDetections: SignalDetectionsState;
  systemMessage: SystemMessageState;
  userSession: UserSessionState;
  waveform: WaveformState;
  workflow: WorkflowState;
}>({
  [analystSlice.name]: analystSlice.reducer,
  [stationPropertiesConfigurationSlice.name]: stationPropertiesConfigurationSlice.reducer,
  [commonSlice.name]: commonSlice.reducer,
  [dataAcquisitionSlice.name]: dataAcquisitionSlice.reducer,
  [eventsSlice.name]: eventsSlice.reducer,
  [mapSlice.name]: mapSlice.reducer,
  [signalDetectionsSlice.name]: signalDetectionsSlice.reducer,
  [systemMessageSlice.name]: systemMessageSlice.reducer,
  [userSessionSlice.name]: userSessionSlice.reducer,
  [waveformSlice.name]: waveformSlice.reducer,
  [workflowSlice.name]: workflowSlice.reducer
});

// Infer the `State` type from the reducer itself
type State = ReturnType<typeof reducers>;

/**
 * The app initial state
 */
export const initialState: State = {
  [analystSlice.name]: analystInitialState,
  [stationPropertiesConfigurationSlice.name]: stationPropertiesConfigurationInitialState,
  [commonSlice.name]: commonInitialState,
  [dataAcquisitionSlice.name]: dataAcquisitionInitialState,
  [eventsSlice.name]: eventsInitialState,
  [mapSlice.name]: mapInitialState,
  [signalDetectionsSlice.name]: signalDetectionsInitialState,
  [systemMessageSlice.name]: systemMessageInitialState,
  [userSessionSlice.name]: userSessionInitialState,
  [waveformSlice.name]: waveformInitialState,
  [workflowSlice.name]: workflowInitialState
};

/**
 * The application state reducer
 *
 * Type inference here was causing errors (it was attempting to export types from third party packages).
 * This caused build errors, which is likely an issue with RTK, immer, or typescript itself. Explicitly
 * typing it here seems to make typescript happy.
 */
export const reducer: ReducerWithInitialState<CombinedState<{
  analyst: AnalystState;
  stationPropertiesConfiguration: StationPropertiesConfigurationState;
  common: CommonState;
  dataAcquisition: DataAcquisitionState;
  events: EventsState;
  map: MapState;
  signalDetections: SignalDetectionsState;
  systemMessage: SystemMessageState;
  userSession: UserSessionState;
  waveform: WaveformState;
  workflow: WorkflowState;
}>> = createReducer(initialState, builder => {
  builder
    .addCase(reset, state => {
      state.analyst = initialState.analyst;
      state.stationPropertiesConfiguration = initialState.stationPropertiesConfiguration;
      state.common = initialState.common;
      state.dataAcquisition = initialState.dataAcquisition;
      state.events = initialState.events;
      state.map = initialState.map;
      state.signalDetections = initialState.signalDetections;
      state.systemMessage = initialState.systemMessage;
      state.userSession = initialState.userSession;
      state.waveform = initialState.waveform;
      state.workflow = initialState.workflow;
    })
    .addDefaultCase(reducers);
});
