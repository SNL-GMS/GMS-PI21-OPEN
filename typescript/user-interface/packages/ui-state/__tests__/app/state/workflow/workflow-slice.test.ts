import type { CommonTypes } from '@gms/common-model';
import { WorkflowTypes } from '@gms/common-model';
import { createAction } from '@reduxjs/toolkit';
import type * as Redux from 'redux';
import type { MockStore, MockStoreCreator } from 'redux-mock-store';
import createMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { initialState } from '../../../../src/ts/app/state/reducer';
import {
  workflowInitialState,
  workflowSlice
} from '../../../../src/ts/app/state/workflow/workflow-slice';
import type { AppState } from '../../../../src/ts/app/store';

const middlewares = [thunk];
const mockStoreCreator: MockStoreCreator<AppState, Redux.AnyAction> = createMockStore(middlewares);
let store: MockStore<AppState, Redux.AnyAction>;

describe('workflow slice', () => {
  describe('exists', () => {
    it('exists', () => {
      expect(workflowInitialState).toBeDefined();
      expect(workflowSlice).toBeDefined();
      expect(workflowInitialState).toMatchSnapshot();
    });
  });

  describe('reducer', () => {
    it('should return the initial state', () => {
      expect(workflowSlice.reducer(undefined, createAction(undefined))).toEqual(
        workflowInitialState
      );
      expect(workflowSlice.reducer(undefined, createAction(''))).toEqual(workflowInitialState);
      expect(workflowSlice.reducer(workflowInitialState, createAction(undefined))).toEqual(
        workflowInitialState
      );
      expect(workflowSlice.reducer(workflowInitialState, createAction(''))).toEqual(
        workflowInitialState
      );
    });
  });
  describe('actions', () => {
    beforeEach(() => {
      store = mockStoreCreator({ app: initialState } as any);
    });

    it('should set the time range', () => {
      const timeRange: CommonTypes.TimeRange = {
        startTimeSecs: 1,
        endTimeSecs: 2
      };
      const action: Redux.AnyAction = {
        type: workflowSlice.actions.setTimeRange.type,
        payload: timeRange
      };

      expect(workflowSlice.reducer(workflowInitialState, action)).toEqual({
        ...workflowInitialState,
        timeRange
      });

      expect(workflowSlice.actions.setTimeRange(timeRange)).toEqual(action);

      store.dispatch(workflowSlice.actions.setTimeRange(timeRange));
      const actions = store.getActions();
      expect(actions).toEqual([action]);
    });

    it('should set the station group', () => {
      const stationGroup: WorkflowTypes.StationGroup = {
        effectiveAt: 1,
        name: 'name',
        description: 'description'
      };
      const action: Redux.AnyAction = {
        type: workflowSlice.actions.setStationGroup.type,
        payload: stationGroup
      };

      expect(workflowSlice.reducer(workflowInitialState, action)).toEqual({
        ...workflowInitialState,
        stationGroup
      });

      expect(workflowSlice.actions.setStationGroup(stationGroup)).toEqual(action);

      store.dispatch(workflowSlice.actions.setStationGroup(stationGroup));
      const actions = store.getActions();
      expect(actions).toEqual([action]);
    });

    it('should set the openIntervalName', () => {
      const openIntervalName = '1';
      const action: Redux.AnyAction = {
        type: workflowSlice.actions.setOpenIntervalName.type,
        payload: openIntervalName
      };

      expect(workflowSlice.reducer(workflowInitialState, action)).toEqual({
        ...workflowInitialState,
        openIntervalName
      });

      expect(workflowSlice.actions.setOpenIntervalName(openIntervalName)).toEqual(action);

      store.dispatch(workflowSlice.actions.setOpenIntervalName(openIntervalName));
      const actions = store.getActions();
      expect(actions).toEqual([action]);
    });

    it('should set the openActivityNames', () => {
      const openActivityNames = ['1'];
      const action: Redux.AnyAction = {
        type: workflowSlice.actions.setOpenActivityNames.type,
        payload: openActivityNames
      };

      expect(workflowSlice.reducer(workflowInitialState, action)).toEqual({
        ...workflowInitialState,
        openActivityNames
      });

      expect(workflowSlice.actions.setOpenActivityNames(openActivityNames)).toEqual(action);

      store.dispatch(workflowSlice.actions.setOpenActivityNames(openActivityNames));
      const actions = store.getActions();
      expect(actions).toEqual([action]);
    });

    it('should set the analysis modes', () => {
      const analysisMode = WorkflowTypes.AnalysisMode.EVENT_REVIEW;
      const action: Redux.AnyAction = {
        type: workflowSlice.actions.setAnalysisMode.type,
        payload: analysisMode
      };

      expect(workflowSlice.reducer(workflowInitialState, action)).toEqual({
        ...workflowInitialState,
        analysisMode
      });

      expect(workflowSlice.actions.setAnalysisMode(analysisMode)).toEqual(action);

      store.dispatch(workflowSlice.actions.setAnalysisMode(analysisMode));
      const actions = store.getActions();
      expect(actions).toEqual([action]);
    });
  });
});
