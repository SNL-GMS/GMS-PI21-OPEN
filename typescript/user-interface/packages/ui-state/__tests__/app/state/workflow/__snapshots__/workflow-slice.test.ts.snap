// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workflow slice exists exists 1`] = `
Object {
  "analysisMode": null,
  "openActivityNames": Array [],
  "openIntervalName": null,
  "stationGroup": null,
  "timeRange": Object {
    "endTimeSecs": null,
    "startTimeSecs": null,
  },
}
`;
